### PR TITLE
Add mode='fan_geo_avg' to nn.initializers.variance_scaling.

### LINKED ISSUE #25649: Add mode='fan_geo_avg' to nn.initializers.variance_scaling
**Feature request:** Add `'fan_geo_avg'` as an option for the `mode` argument of [`nn.initializers.variance_scaling`](https://jax.readthedocs.io/en/latest/_autosummary/jax.nn.initializers.variance_scaling.html), in order to use the [geometric mean](https://en.wikipedia.org/wiki/Geometric_mean) rather than [arithmetic mean](https://en.wikipedia.org/wiki/Arithmetic_mean) of `fan_in` and `fan_out` for the denominator.

[Beyond Folklore: A Scaling Calculus for the Design and Initialization of ReLU Networks](https://arxiv.org/abs/1906.04267):

> This scaling calculus results in a number of consequences, among them the fact that the **geometric mean** of the fan-in and fan-out, rather than the fan-in, fan-out, or arithmetic mean, should be used for the initialization of the variance of weights in a neural network.
>
> Initialization using the **geometric-mean** of the fan-in and fan-out ensures a constant layer scaling factor throughout the network, aiding optimization.
>
> The use of **geometric initialization** results in an equally weighted diagonal, in contrast to the other initializations considered.

[SplitNets: Designing Neural Architectures for Efficient Distributed Computing on Head-Mounted Systems](https://arxiv.org/abs/2204.04705):

> Using **geometric average** allows us to find a better compromise between forward and backward passes and significantly improve training stability and final accuracy
>
> Our split-aware initialization adopts **geometric average** instead of arithmetic average to make a better balance between forward and backward

I can submit a PR for this.

### PR DESCRIPTION
**Pull request by carlosgmartin**
Fixes #25649.

### CODE REVIEW

**Comment by carlosgmartin** [comment]
@jakevdp How does this look?

**Review by jakevdp — COMMENTED** [review]
Looks reasonable. Can we add a test?

**Comment by carlosgmartin** [comment]
@jakevdp Done.

**Comment by jakevdp** [comment]
Wow, we have surprisingly little test coverage for `jax.nn`.

**Review by jakevdp — APPROVED** [review]
